node/src/components/rpc_server/rpcs/state.rs

```diff
@@ -929,6 +929,155 @@ impl RpcWithParamsExt for QueryGlobalState {
+        effect_builder: EffectBuilder<REv>,
+        response_builder: Builder,
+        params: Self::RequestParams,
+        api_version: ProtocolVersion,
+    ) -> BoxFuture<'static, Result<Response<Body>, Error>> {
+        async move {
+            let state_root_hash = match params.state_identifier {
+                None => match effect_builder.get_highest_block_header_from_storage().await {
+                    None => {
+                        return Ok(response_builder.error(warp_json_rpc::Error::custom(
+                            ErrorCode::NoSuchBlock as i64,
+                            "query-balance failed to retrieve highest block header",
+                        ))?)
+                    }
+                    Some(block_header) => *block_header.state_root_hash(),
+                },
+                Some(state_identifier) => {
+                    let (state_root_hash, _) = match get_state_root_hash_and_optional_header(
+                        effect_builder,
+                        state_identifier,
+                        Self::METHOD,
+                    )
+                    .await
+                    {
+                        Ok(state_root_hash) => state_root_hash,
+                        Err(error) => return Ok(response_builder.error(error)?),
+                    };
+                    state_root_hash
+                }
+            };
+
+            let purse_uref = match params.balance_identifier {
+                PurseIdentifier::MainPurseUnderPublicKey(account_public_key) => {
+                    match get_account(
+                        effect_builder,
+                        state_root_hash,
+                        account_public_key.to_account_hash(),
+                    )
+                    .await
+                    {
+                        Ok(account) => account.main_purse(),
+                        Err(error) => return Ok(response_builder.error(error)?),
+                    }
+                }
+                PurseIdentifier::MainPurseUnderAccountHash(account_hash) => {
+                    match get_account(effect_builder, state_root_hash, account_hash).await {
+                        Ok(account) => account.main_purse(),
+                        Err(error) => return Ok(response_builder.error(error)?),
+                    }
+                }
+                PurseIdentifier::PurseUref(purse_uref) => purse_uref,
+            };
+
+            // Get the balance.
+            let balance_result = effect_builder
+                .make_request(
+                    |responder| RpcRequest::GetBalance {
+                        state_root_hash,
+                        purse_uref,
+                        responder,
+                    },
+                    QueueKind::Api,
+                )
+                .await;
+
+            let balance_value = match balance_result {
+                Ok(BalanceResult::Success { motes, .. }) => motes,
+                Ok(balance_result) => {
```

Comment:
```suggestion
                Ok(BalanceResult::RootNotFound) => {
```

---

**Rationale:**
The `Ok(balance_result)` is a bit like a "catch all others" pattern, similar to `_`. In case someone adds a new variant to `BalanceResult` it will still compile assuming that this new variant is an error (!Success), but the new variant could be as well another flavor of `Success`, which should be handled correctly.

Reply:
Applied suggestion